pkg/kubectl/cmd/util/helpers.go

```diff
@@ -436,11 +442,45 @@ func Merge(codec runtime.Codec, dst runtime.Object, fragment string) (runtime.Ob
 	if err != nil {
 		return nil, err
 	}
+
 	patched, err := jsonpatch.MergePatch(target, []byte(fragment))
 	if err != nil {
 		return nil, err
 	}
-	out, err := runtime.Decode(codec, patched)
+
+	/* Get API version and kind */
+	var tMeta metav1.TypeMeta
+	if err := json.Unmarshal([]byte(fragment), &tMeta); err != nil {
+		return nil, err
+	}
+
+	/* Get the modified object */
+	var modObject  runtime.Object
+	switch  dst.(type) {
+	case *v1.Pod:
+		modObject = &v1.Pod{}
+	case *v1.Service:
+		modObject = &v1.Service{}
+	case *v1.ReplicationController:
+		modObject = &v1.ReplicationController{}
+	case *appsv1.Deployment:
+		modObject = &appsv1.Deployment{}
+	case *appsv1.ReplicaSet:
+		modObject = &appsv1.ReplicaSet{}
+	case *extensionsv1beta1.Deployment:
+		modObject = &extensionsv1beta1.Deployment{}
+	case *extensionsv1beta1.ReplicaSet:
+		modObject = &extensionsv1beta1.ReplicaSet{}
+	default:
+		return nil, errors.New("Doesn't support the expose type")
+	}
+
+	if err := json.Unmarshal([]byte(patched), &modObject); err != nil {
+		return nil, err
+	}
+
+	out, err := legacyscheme.Scheme.ConvertToVersion(modObject,
+			schema.FromAPIVersionAndKind(tMeta.APIVersion, tMeta.Kind).GroupVersion())
```

Comment:
> If the user doesn't specify apiVersion or kind in the commandline, it would be a problem.

how so? if the patch doesn't specify apiVersion or kind, the original object's apiVersion/kind is preserved:

```
kubectl create -f test/fixtures/pkg/kubectl/cmd/expose/appsv1deployment.yaml

kubectl expose -f test/fixtures/pkg/kubectl/cmd/expose/appsv1deployment.yaml \
  --dry-run -o yaml \
  --overrides='{"spec":{"selector":{"foo":"bar"}}}'

apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    name: expose-test-deployment
  name: expose-test-deployment
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  selector:
    foo: bar
    name: nginx
status:
  loadBalancer: {}
```